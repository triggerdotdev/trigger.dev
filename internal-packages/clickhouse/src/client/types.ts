import type { Result } from "@trigger.dev/core/v3";
import type { z } from "zod";
import type { InsertError, QueryError } from "./errors.js";
import { ClickHouseSettings } from "@clickhouse/client";
import type { BaseQueryParams, InsertResult } from "@clickhouse/client";

export type ClickhouseQueryFunction<TInput, TOutput> = (
  params: TInput,
  options?: {
    attributes?: Record<string, string | number | boolean>;
    params?: BaseQueryParams;
  }
) => Promise<Result<TOutput[], QueryError>>;

export interface ClickhouseReader {
  query<TIn extends z.ZodSchema<any>, TOut extends z.ZodSchema<any>>(req: {
    /**
     * The name of the operation.
     * This will be used to identify the operation in the span.
     */
    name: string;
    /**
     * The SQL query to run.
     * Use {paramName: Type} to define parameters
     * Example: `SELECT * FROM table WHERE id = {id: String}`
     */
    query: string;
    /**
     * The schema of the parameters
     * Example: z.object({ id: z.string() })
     */
    params?: TIn;
    /**
     * The schema of the output of each row
     * Example: z.object({ id: z.string() })
     */
    schema: TOut;
    /**
     * The settings to use for the query.
     * These will be merged with the default settings.
     */
    settings?: ClickHouseSettings;
  }): ClickhouseQueryFunction<z.input<TIn>, z.output<TOut>>;

  close(): Promise<void>;
}

export type ClickhouseInsertFunction<TInput> = (
  events: TInput | TInput[],
  options?: {
    attributes?: Record<string, string | number | boolean>;
    params?: BaseQueryParams;
  }
) => Promise<Result<InsertResult, InsertError>>;

export interface ClickhouseWriter {
  insert<TSchema extends z.ZodSchema<any>>(req: {
    name: string;
    table: string;
    schema: TSchema;
    settings?: ClickHouseSettings;
  }): ClickhouseInsertFunction<z.input<TSchema>>;

  close(): Promise<void>;
}
