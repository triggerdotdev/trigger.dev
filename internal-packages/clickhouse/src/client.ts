import { createClient as createClickhouseClient } from "@clickhouse/client";

export function createClient(url: string) {
  const client = createClickhouseClient({
    url,
  });

  return client;
}
