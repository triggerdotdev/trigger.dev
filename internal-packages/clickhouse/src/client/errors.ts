export type ErrorContext = Record<string, unknown>;

export abstract class BaseError<TContext extends ErrorContext = ErrorContext> extends Error {
  public abstract readonly retry: boolean;
  public readonly cause: BaseError | undefined;
  public readonly context: TContext | undefined;
  public readonly message: string;
  public abstract readonly name: string;

  constructor(opts: { message: string; cause?: BaseError; context?: TContext }) {
    super(opts.message);
    this.message = opts.message;
    this.cause = opts.cause;
    this.context = opts.context;
  }

  public toString(): string {
    return `${this.name}: ${this.message} - ${JSON.stringify(
      this.context
    )} - caused by ${this.cause?.toString()}`;
  }
}

export class InsertError extends BaseError {
  public readonly retry = true;
  public readonly name = InsertError.name;
  constructor(message: string) {
    super({
      message,
    });
  }
}
export class QueryError extends BaseError<{ query: string }> {
  public readonly retry = true;
  public readonly name = QueryError.name;
  constructor(message: string, context: { query: string }) {
    super({
      message,
      context,
    });
  }
}
