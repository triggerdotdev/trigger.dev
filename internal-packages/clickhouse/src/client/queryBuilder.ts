import { z } from "zod";
import { ClickhouseQueryFunction, ClickhouseReader } from "./types.js";
import { ClickHouseSettings } from "@clickhouse/client";
export type QueryParamValue = string | number | boolean | Array<string | number | boolean> | null;
export type QueryParams = Record<string, QueryParamValue>;

export class ClickhouseQueryBuilder<TOutput> {
  private name: string;
  private baseQuery: string;
  private whereClauses: string[] = [];
  private params: QueryParams = {};
  private orderByClause: string | null = null;
  private limitClause: string | null = null;
  private reader: ClickhouseReader;
  private schema: z.ZodSchema<TOutput>;
  private settings: ClickHouseSettings | undefined;
  private groupByClause: string | null = null;

  constructor(
    name: string,
    baseQuery: string,
    reader: ClickhouseReader,
    schema: z.ZodSchema<TOutput>,
    settings?: ClickHouseSettings
  ) {
    this.name = name;
    this.baseQuery = baseQuery;
    this.reader = reader;
    this.schema = schema;
    this.settings = settings;
  }

  where(clause: string, params?: QueryParams): this {
    this.whereClauses.push(clause);
    if (params) {
      Object.assign(this.params, params);
    }
    return this;
  }

  whereIf(condition: any, clause: string, params?: QueryParams): this {
    if (condition) {
      this.where(clause, params);
    }
    return this;
  }

  groupBy(clause: string): this {
    this.groupByClause = clause;
    return this;
  }

  orderBy(clause: string): this {
    this.orderByClause = clause;
    return this;
  }

  limit(limit: number): this {
    this.limitClause = `LIMIT ${limit}`;
    return this;
  }

  execute(): ReturnType<ClickhouseQueryFunction<void, TOutput>> {
    const { query, params } = this.build();

    const queryFunction = this.reader.query({
      name: this.name,
      query,
      params: z.any(),
      schema: this.schema,
      settings: this.settings,
    });

    return queryFunction(params);
  }

  build(): { query: string; params: QueryParams } {
    let query = this.baseQuery;
    if (this.whereClauses.length > 0) {
      query += " WHERE " + this.whereClauses.join(" AND ");
    }
    if (this.groupByClause) {
      query += ` GROUP BY ${this.groupByClause}`;
    }
    if (this.orderByClause) {
      query += ` ORDER BY ${this.orderByClause}`;
    }
    if (this.limitClause) {
      query += ` ${this.limitClause}`;
    }
    return { query, params: this.params };
  }
}
