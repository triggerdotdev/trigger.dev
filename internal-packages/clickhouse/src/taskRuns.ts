import { ClickHouseSettings } from "@clickhouse/client";
import { z } from "zod";
import { ClickhouseWriter } from "./client/types.js";

export const TaskRunV1 = z.object({
  environment_id: z.string(),
  organization_id: z.string(),
  project_id: z.string(),
  run_id: z.string(),
  updated_at: z.number().int(),
  created_at: z.number().int(),
  status: z.string(),
  environment_type: z.string(),
  friendly_id: z.string(),
  attempt: z.number().int().default(1),
  engine: z.string(),
  task_identifier: z.string(),
  queue: z.string(),
  schedule_id: z.string(),
  batch_id: z.string(),
  completed_at: z.number().int().nullish(),
  started_at: z.number().int().nullish(),
  executed_at: z.number().int().nullish(),
  delay_until: z.number().int().nullish(),
  queued_at: z.number().int().nullish(),
  expired_at: z.number().int().nullish(),
  usage_duration_ms: z.number().int().default(0),
  cost_in_cents: z.number().default(0),
  base_cost_in_cents: z.number().default(0),
  output: z.unknown(),
  error: z.unknown(),
  tags: z.array(z.string()).default([]),
  task_version: z.string(),
  sdk_version: z.string(),
  cli_version: z.string(),
  machine_preset: z.string(),
  root_run_id: z.string(),
  parent_run_id: z.string(),
  depth: z.number().int().default(0),
  span_id: z.string(),
  trace_id: z.string(),
  idempotency_key: z.string(),
  expiration_ttl: z.string(),
  is_test: z.boolean().default(false),
  _version: z.string(),
  _is_deleted: z.number().int().default(0),
});

export type TaskRunV1 = z.input<typeof TaskRunV1>;

export function insertTaskRuns(ch: ClickhouseWriter, settings?: ClickHouseSettings) {
  return ch.insert({
    name: "insertTaskRuns",
    table: "trigger_dev.task_runs_v1",
    schema: TaskRunV1,
    settings: {
      async_insert: 1,
      wait_for_async_insert: 0,
      async_insert_max_data_size: "1000000",
      async_insert_busy_timeout_ms: 1000,
      enable_json_type: 1,
      ...settings,
    },
  });
}

export const RawTaskRunPayloadV1 = z.object({
  run_id: z.string(),
  created_at: z.number().int(),
  payload: z.unknown(),
});

export type RawTaskRunPayloadV1 = z.infer<typeof RawTaskRunPayloadV1>;

export function insertRawTaskRunPayloads(ch: ClickhouseWriter, settings?: ClickHouseSettings) {
  return ch.insert({
    name: "insertRawTaskRunPayloads",
    table: "trigger_dev.raw_task_runs_payload_v1",
    schema: RawTaskRunPayloadV1,
    settings: {
      async_insert: 1,
      wait_for_async_insert: 0,
      async_insert_max_data_size: "1000000",
      async_insert_busy_timeout_ms: 1000,
      enable_json_type: 1,
      ...settings,
    },
  });
}
