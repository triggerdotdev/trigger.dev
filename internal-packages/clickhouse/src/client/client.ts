import {
  type ClickHouseClient,
  ClickHouseError,
  type ClickHouseSettings,
  createClient,
} from "@clickhouse/client";
import { recordSpanError, Span, startSpan, trace, Tracer } from "@internal/tracing";
import { flattenAttributes, tryCatch } from "@trigger.dev/core/v3";
import { z } from "zod";
import { InsertError, QueryError } from "./errors.js";
import type {
  ClickhouseInsertFunction,
  ClickhouseQueryFunction,
  ClickhouseReader,
  ClickhouseWriter,
} from "./types.js";
import { generateErrorMessage } from "zod-error";
import { Logger } from "@trigger.dev/core/logger";

export type ClickhouseConfig = {
  name: string;
  url: string;
  tracer?: Tracer;
  clickhouseSettings?: ClickHouseSettings;
  logger?: Logger;
};

export class ClickhouseClient implements ClickhouseReader, ClickhouseWriter {
  public readonly client: ClickHouseClient;
  private readonly tracer: Tracer;
  private readonly name: string;
  private readonly logger: Logger;

  constructor(config: ClickhouseConfig) {
    this.name = config.name;
    this.logger = config.logger ?? new Logger("ClickhouseClient", "debug");

    this.client = createClient({
      url: config.url,

      clickhouse_settings: {
        ...config.clickhouseSettings,
        output_format_json_quote_64bit_integers: 0,
        output_format_json_quote_64bit_floats: 0,
      },
    });

    this.tracer = config.tracer ?? trace.getTracer("@internal/clickhouse");
  }

  public async close() {
    await this.client.close();
  }

  public query<TIn extends z.ZodSchema<any>, TOut extends z.ZodSchema<any>>(req: {
    /**
     * The name of the operation.
     * This will be used to identify the operation in the span.
     */
    name: string;
    /**
     * The SQL query to run.
     * Use {paramName: Type} to define parameters
     * Example: `SELECT * FROM table WHERE id = {id: String}`
     */
    query: string;
    /**
     * The schema of the parameters
     * Example: z.object({ id: z.string() })
     */
    params?: TIn;
    /**
     * The schema of the output of each row
     * Example: z.object({ id: z.string() })
     */
    schema: TOut;
    /**
     * The settings to use for the query.
     * These will be merged with the default settings.
     */
    settings?: ClickHouseSettings;
  }): ClickhouseQueryFunction<z.input<TIn>, z.output<TOut>> {
    return async (params, options) => {
      return await startSpan(this.tracer, "query", async (span) => {
        span.setAttributes({
          "clickhouse.clientName": this.name,
          "clickhouse.operationName": req.name,
          ...flattenAttributes(req.settings, "clickhouse.settings"),
          ...flattenAttributes(options?.attributes),
        });

        const validParams = req.params?.safeParse(params);

        if (validParams?.error) {
          recordSpanError(span, validParams.error);

          this.logger.error("Error parsing query params", {
            name: req.name,
            error: validParams.error,
            query: req.query,
            params,
          });

          return [
            new QueryError(`Bad params: ${generateErrorMessage(validParams.error.issues)}`, {
              query: req.query,
            }),
            null,
          ];
        }

        let unparsedRows: Array<TOut> = [];

        const [clickhouseError, res] = await tryCatch(
          this.client.query({
            query: req.query,
            query_params: validParams?.data,
            format: "JSONEachRow",
            ...options?.params,
            clickhouse_settings: {
              ...req.settings,
              ...options?.params?.clickhouse_settings,
            },
          })
        );

        if (clickhouseError) {
          this.logger.error("Error querying clickhouse", {
            name: req.name,
            error: clickhouseError,
            query: req.query,
            params,
          });

          recordClickhouseError(span, clickhouseError);

          return [
            new QueryError(`Unable to query clickhouse: ${clickhouseError.message}`, {
              query: req.query,
            }),
            null,
          ];
        }

        unparsedRows = await res.json();

        span.setAttributes({
          "clickhouse.query_id": res.query_id,
          ...flattenAttributes(res.response_headers, "clickhouse.response_headers"),
        });

        const summaryHeader = res.response_headers["x-clickhouse-summary"];

        if (typeof summaryHeader === "string") {
          span.setAttributes({
            ...flattenAttributes(JSON.parse(summaryHeader), "clickhouse.summary"),
          });
        }

        const parsed = z.array(req.schema).safeParse(unparsedRows);

        if (parsed.error) {
          this.logger.error("Error parsing clickhouse query result", {
            name: req.name,
            error: parsed.error,
            query: req.query,
            params,
          });

          const queryError = new QueryError(generateErrorMessage(parsed.error.issues), {
            query: req.query,
          });

          recordSpanError(span, queryError);

          return [queryError, null];
        }

        span.setAttributes({
          "clickhouse.rows": unparsedRows.length,
        });

        return [null, parsed.data];
      });
    };
  }

  public insert<TSchema extends z.ZodSchema<any>>(req: {
    name: string;
    table: string;
    schema: TSchema;
    settings?: ClickHouseSettings;
  }): ClickhouseInsertFunction<z.input<TSchema>> {
    return async (events, options) => {
      return await startSpan(this.tracer, "insert", async (span) => {
        span.setAttributes({
          "clickhouse.clientName": this.name,
          "clickhouse.tableName": req.table,
          "clickhouse.operationName": req.name,
          ...flattenAttributes(req.settings, "clickhouse.settings"),
          ...flattenAttributes(options?.attributes),
        });

        let validatedEvents: z.output<TSchema> | z.output<TSchema>[] | undefined = undefined;

        const v = Array.isArray(events)
          ? req.schema.array().safeParse(events)
          : req.schema.safeParse(events);

        if (!v.success) {
          this.logger.error("Error validating insert events", {
            name: req.name,
            table: req.table,
            error: v.error,
          });

          const error = new InsertError(generateErrorMessage(v.error.issues));

          recordSpanError(span, error);

          return [error, null];
        }

        validatedEvents = v.data;

        const [clickhouseError, result] = await tryCatch(
          this.client.insert({
            table: req.table,
            format: "JSONEachRow",
            values: Array.isArray(validatedEvents) ? validatedEvents : [validatedEvents],
            ...options?.params,
            clickhouse_settings: {
              ...req.settings,
              ...options?.params?.clickhouse_settings,
            },
          })
        );

        if (clickhouseError) {
          this.logger.error("Error inserting into clickhouse", {
            name: req.name,
            error: clickhouseError,
            table: req.table,
          });

          recordClickhouseError(span, clickhouseError);

          return [new InsertError(clickhouseError.message), null];
        }

        span.setAttributes({
          "clickhouse.query_id": result.query_id,
          "clickhouse.executed": result.executed,
          "clickhouse.summary.read_rows": result.summary?.read_rows,
          "clickhouse.summary.read_bytes": result.summary?.read_bytes,
          "clickhouse.summary.written_rows": result.summary?.written_rows,
          "clickhouse.summary.written_bytes": result.summary?.written_bytes,
          "clickhouse.summary.total_rows_to_read": result.summary?.total_rows_to_read,
          "clickhouse.summary.result_rows": result.summary?.result_rows,
          "clickhouse.summary.result_bytes": result.summary?.result_bytes,
          "clickhouse.summary.elapsed_ns": result.summary?.elapsed_ns,
        });

        return [null, result];
      });
    };
  }
}

function recordClickhouseError(span: Span, error: Error) {
  if (error instanceof ClickHouseError) {
    span.setAttributes({
      "clickhouse.error.code": error.code,
      "clickhouse.error.message": error.message,
      "clickhouse.error.type": error.type,
    });
    recordSpanError(span, error);
  } else {
    recordSpanError(span, error);
  }
}
