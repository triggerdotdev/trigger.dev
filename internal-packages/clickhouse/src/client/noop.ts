import { Result } from "@trigger.dev/core/v3";
import { InsertError, QueryError } from "./errors.js";
import { ClickhouseWriter } from "./types.js";
import { ClickhouseReader } from "./types.js";
import { z } from "zod";
import { ClickHouseSettings, InsertResult } from "@clickhouse/client";

export class NoopClient implements ClickhouseReader, ClickhouseWriter {
  public async close() {
    return;
  }

  public query<TIn extends z.ZodSchema<any>, TOut extends z.ZodSchema<any>>(req: {
    query: string;
    params?: TIn;
    schema: TOut;
  }): (params: z.input<TIn>) => Promise<Result<z.output<TOut>[], QueryError>> {
    return async (params: z.input<TIn>) => {
      const validParams = req.params?.safeParse(params);

      if (validParams?.error) {
        return [new QueryError(`Bad params: ${validParams.error.message}`, { query: "" }), null];
      }

      return [null, []];
    };
  }

  public insert<TSchema extends z.ZodSchema<any>>(req: {
    name: string;
    table: string;
    schema: TSchema;
    settings?: ClickHouseSettings;
  }): (
    events: z.input<TSchema> | z.input<TSchema>[]
  ) => Promise<Result<InsertResult, InsertError>> {
    return async (events: z.input<TSchema> | z.input<TSchema>[]) => {
      const v = Array.isArray(events)
        ? req.schema.array().safeParse(events)
        : req.schema.safeParse(events);

      if (!v.success) {
        return [new InsertError(v.error.message), null];
      }

      return [
        null,
        {
          executed: true,
          query_id: "noop",
          summary: {
            read_rows: "0",
            read_bytes: "0",
            written_rows: "0",
            written_bytes: "0",
            total_rows_to_read: "0",
            result_rows: "0",
            result_bytes: "0",
            elapsed_ns: "0",
          },
          response_headers: {},
        },
      ];
    };
  }
}
