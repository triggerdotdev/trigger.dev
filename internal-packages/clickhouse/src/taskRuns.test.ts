import { clickhouseTest } from "@internal/testcontainers";
import { z } from "zod";
import { ClickhouseClient } from "./client/client.js";
import { insertRawTaskRunPayloads, insertTaskRuns } from "./taskRuns.js";

describe("Task Runs V1", () => {
  clickhouseTest("should be able to insert task runs", async ({ clickhouseContainer }) => {
    const client = new ClickhouseClient({
      name: "test",
      url: clickhouseContainer.getConnectionUrl(),
    });

    const insert = insertTaskRuns(client, {
      async_insert: 0, // turn off async insert for this test
    });

    const insertPayloads = insertRawTaskRunPayloads(client, {
      async_insert: 0, // turn off async insert for this test
    });

    const [insertError, insertResult] = await insert([
      {
        environment_id: "env_1234",
        environment_type: "DEVELOPMENT",
        organization_id: "org_1234",
        project_id: "project_1234",
        run_id: "run_1234",
        friendly_id: "friendly_1234",
        attempt: 1,
        engine: "V2",
        status: "PENDING",
        task_identifier: "my-task",
        queue: "my-queue",
        schedule_id: "schedule_1234",
        batch_id: "batch_1234",
        created_at: Date.now(),
        updated_at: Date.now(),
        completed_at: undefined,
        tags: ["tag1", "tag2"],
        output: {
          key: "value",
        },
        error: {
          type: "BUILT_IN_ERROR",
          name: "Error",
          message: "error",
          stackTrace: "stack trace",
        },
        usage_duration_ms: 1000,
        cost_in_cents: 100,
        task_version: "1.0.0",
        sdk_version: "1.0.0",
        cli_version: "1.0.0",
        machine_preset: "small-1x",
        is_test: true,
        span_id: "span_1234",
        trace_id: "trace_1234",
        idempotency_key: "idempotency_key_1234",
        expiration_ttl: "1h",
        root_run_id: "root_run_1234",
        parent_run_id: "parent_run_1234",
        depth: 1,
        _version: "1",
      },
    ]);

    expect(insertError).toBeNull();
    expect(insertResult).toEqual(expect.objectContaining({ executed: true }));
    expect(insertResult?.summary?.written_rows).toEqual("1");

    const query = client.query({
      name: "query-task-runs",
      query: "SELECT * FROM trigger_dev.task_runs_v1",
      schema: z.object({
        environment_id: z.string(),
        run_id: z.string(),
      }),
      params: z.object({
        run_id: z.string(),
      }),
    });

    const [queryError, result] = await query({ run_id: "run_1234" });

    expect(queryError).toBeNull();
    expect(result).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          environment_id: "env_1234",
          run_id: "run_1234",
        }),
      ])
    );

    const [insertPayloadsError, insertPayloadsResult] = await insertPayloads([
      {
        run_id: "run_1234",
        created_at: Date.now(),
        payload: {
          key: "value",
        },
      },
    ]);

    expect(insertPayloadsError).toBeNull();
    expect(insertPayloadsResult).toEqual(expect.objectContaining({ executed: true }));
    expect(insertPayloadsResult?.summary?.written_rows).toEqual("1");

    const queryPayloads = client.query({
      name: "query-raw-task-run-payloads",
      query: "SELECT * FROM trigger_dev.raw_task_runs_payload_v1",
      schema: z.object({
        run_id: z.string(),
        created_at: z.coerce.date(),
        payload: z.unknown(),
      }),
    });

    const [queryPayloadsError, resultPayloads] = await queryPayloads({ run_id: "run_1234" });

    expect(queryPayloadsError).toBeNull();
    expect(resultPayloads).toEqual(
      expect.arrayContaining([expect.objectContaining({ run_id: "run_1234" })])
    );
  });

  clickhouseTest("should deduplicate on the _version column", async ({ clickhouseContainer }) => {
    const client = new ClickhouseClient({
      name: "test",
      url: clickhouseContainer.getConnectionUrl(),
    });

    const insert = insertTaskRuns(client, {
      async_insert: 0, // turn off async insert for this test
    });

    const [insertError, insertResult] = await insert([
      {
        environment_id: "cm9kddfcs01zqdy88ld9mmrli",
        organization_id: "cm8zs78wb0002dy616dg75tv3",
        project_id: "cm9kddfbz01zpdy88t9dstecu",
        run_id: "cma45oli70002qrdy47w0j4n7",
        environment_type: "PRODUCTION",
        friendly_id: "run_cma45oli70002qrdy47w0j4n7",
        attempt: 1,
        engine: "V2",
        status: "PENDING",
        task_identifier: "retry-task",
        queue: "task/retry-task",
        schedule_id: "",
        batch_id: "",
        root_run_id: "",
        parent_run_id: "",
        depth: 0,
        span_id: "538677637f937f54",
        trace_id: "20a28486b0b9f50c647b35e8863e36a5",
        idempotency_key: "",
        created_at: new Date("2025-04-30 16:34:04.312").getTime(),
        updated_at: new Date("2025-04-30 16:34:04.312").getTime(),
        started_at: null,
        executed_at: null,
        completed_at: null,
        delay_until: null,
        queued_at: new Date("2025-04-30 16:34:04.311").getTime(),
        expired_at: null,
        expiration_ttl: "",
        usage_duration_ms: 0,
        cost_in_cents: 0,
        base_cost_in_cents: 0,
        output: null,
        error: null,
        tags: [],
        task_version: "",
        sdk_version: "",
        cli_version: "",
        machine_preset: "",
        is_test: true,
        _version: "1",
      },
      {
        environment_id: "cm9kddfcs01zqdy88ld9mmrli",
        organization_id: "cm8zs78wb0002dy616dg75tv3",
        project_id: "cm9kddfbz01zpdy88t9dstecu",
        run_id: "cma45oli70002qrdy47w0j4n7",
        environment_type: "PRODUCTION",
        friendly_id: "run_cma45oli70002qrdy47w0j4n7",
        attempt: 1,
        engine: "V2",
        status: "COMPLETED_SUCCESSFULLY",
        task_identifier: "retry-task",
        queue: "task/retry-task",
        schedule_id: "",
        batch_id: "",
        root_run_id: "",
        parent_run_id: "",
        depth: 0,
        span_id: "538677637f937f54",
        trace_id: "20a28486b0b9f50c647b35e8863e36a5",
        idempotency_key: "",
        created_at: new Date("2025-04-30 16:34:04.312").getTime(),
        updated_at: new Date("2025-04-30 16:34:04.312").getTime(),
        started_at: null,
        executed_at: null,
        completed_at: null,
        delay_until: null,
        queued_at: new Date("2025-04-30 16:34:04.311").getTime(),
        expired_at: null,
        expiration_ttl: "",
        usage_duration_ms: 0,
        cost_in_cents: 0,
        base_cost_in_cents: 0,
        output: null,
        error: null,
        tags: [],
        task_version: "",
        sdk_version: "",
        cli_version: "",
        machine_preset: "",
        is_test: true,
        _version: "2",
      },
    ]);

    expect(insertError).toBeNull();
    expect(insertResult).toEqual(expect.objectContaining({ executed: true }));

    const query = client.query({
      name: "query-task-runs",
      query: "SELECT * FROM trigger_dev.task_runs_v1 FINAL",
      schema: z.object({
        environment_id: z.string(),
        run_id: z.string(),
        status: z.string(),
      }),
      params: z.object({
        run_id: z.string(),
      }),
    });

    const [queryError, result] = await query({ run_id: "cma45oli70002qrdy47w0j4n7" });

    expect(queryError).toBeNull();
    expect(result).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          environment_id: "cm9kddfcs01zqdy88ld9mmrli",
          run_id: "cma45oli70002qrdy47w0j4n7",
          status: "COMPLETED_SUCCESSFULLY",
        }),
      ])
    );
  });
});
