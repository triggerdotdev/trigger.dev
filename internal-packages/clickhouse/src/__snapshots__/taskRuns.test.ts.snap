// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task Runs V2 > should be able to query task runs using the query builder 1`] = `
{
  "params": {
    "environmentId": "cm9kddfcs01zqdy88ld9mmrli",
  },
  "query": "SELECT run_id FROM trigger_dev.task_runs_v2 FINAL WHERE environment_id = {environmentId: String}",
}
`;

exports[`Task Runs V2 > should be able to query task runs using the query builder 2`] = `
{
  "params": {
    "environmentId": "cm9kddfcs01zqdy88ld9mmrli",
    "status": "COMPLETED_SUCCESSFULLY",
  },
  "query": "SELECT run_id FROM trigger_dev.task_runs_v2 FINAL WHERE environment_id = {environmentId: String} AND status = {status: String}",
}
`;
